// @hyperlane-xyz/core/dist/factories/contracts/mock/MockERC5164.sol/MockMessageExecutor__factory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MockMessageExecutor__factory = void 0;
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
const ethers_1 = require("ethers");
const _abi = [
    {
        anonymous: false,
        inputs: [
            {
                indexed: true,
                internalType: "uint256",
                name: "fromChainId",
                type: "uint256",
            },
            {
                indexed: true,
                internalType: "bytes32",
                name: "messageId",
                type: "bytes32",
            },
        ],
        name: "MessageIdExecuted",
        type: "event",
    },
];
const _bytecode = "0x6080604052348015600f57600080fd5b50603f80601d6000396000f3fe6080604052600080fdfea26469706673582212200c09666429d18018183ea74eb90f8f4cd30f15af41c2eb093ab30910af69ed9e64736f6c63430008110033";
const isSuperArgs = (xs) => xs.length > 1;
class MockMessageExecutor__factory extends ethers_1.ContractFactory {
    constructor(...args) {
        if (isSuperArgs(args)) {
            super(...args);
        }
        else {
            super(_abi, _bytecode, args[0]);
        }
    }
    deploy(overrides) {
        return super.deploy(overrides || {});
    }
    getDeployTransaction(overrides) {
        return super.getDeployTransaction(overrides || {});
    }
    attach(address) {
        return super.attach(address);
    }
    connect(signer) {
        return super.connect(signer);
    }
    static createInterface() {
        return new ethers_1.utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    }
}
exports.MockMessageExecutor__factory = MockMessageExecutor__factory;
MockMessageExecutor__factory.bytecode = _bytecode;
MockMessageExecutor__factory.abi = _abi;
