// @hyperlane-xyz/sdk/dist/core/TestCoreApp.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TestCoreApp = exports.testCoreFactories = void 0;
const core_1 = require("@hyperlane-xyz/core");
const utils_1 = require("@hyperlane-xyz/utils");
const HyperlaneCore_1 = require("./HyperlaneCore");
const contracts_1 = require("./contracts");
exports.testCoreFactories = {
    ...contracts_1.coreFactories,
    mailbox: new core_1.TestMailbox__factory(),
};
class TestCoreApp extends HyperlaneCore_1.HyperlaneCore {
    getContracts(chain) {
        return super.getContracts(chain);
    }
    async processMessages() {
        const responses = new Map();
        for (const origin of this.chains()) {
            const outbound = await this.processOutboundMessages(origin);
            const originResponses = new Map();
            this.remoteChains(origin).forEach((destination) => originResponses.set(destination, outbound.get(destination)));
            responses.set(origin, originResponses);
        }
        return responses;
    }
    async processOutboundMessages(origin) {
        const responses = new Map();
        const contracts = this.getContracts(origin);
        const outbox = contracts.mailbox;
        const dispatchFilter = outbox.filters.Dispatch();
        const dispatches = await outbox.queryFilter(dispatchFilter);
        for (const dispatch of dispatches) {
            const destination = dispatch.args.destination;
            if (destination === this.multiProvider.getDomainId(origin)) {
                throw new Error('Dispatched message to local domain');
            }
            const destinationChain = this.multiProvider.getChainName(destination);
            const inbox = this.getContracts(destinationChain).mailbox;
            const id = utils_1.utils.messageId(dispatch.args.message);
            const delivered = await inbox.delivered(id);
            if (!delivered) {
                const response = await inbox.process('0x', dispatch.args.message);
                const destinationResponses = responses.get(destinationChain) || [];
                destinationResponses.push(response);
                responses.set(destinationChain, destinationResponses);
            }
        }
        return responses;
    }
}
exports.TestCoreApp = TestCoreApp;
