// @hyperlane-xyz/sdk/dist/utils/wagmi.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.chainMetadataToWagmiChain = exports.wagmiChainMetadata = void 0;
const chainMetadata_1 = require("../consts/chainMetadata");
const objects_1 = require("./objects");
// For convenient use in wagmi-based apps
exports.wagmiChainMetadata = (0, objects_1.objMap)(chainMetadata_1.chainMetadata, (_, metadata) => chainMetadataToWagmiChain(metadata));
function chainMetadataToWagmiChain(metadata) {
    return {
        id: metadata.chainId,
        name: metadata.displayName || metadata.name,
        network: metadata.name,
        nativeCurrency: metadata.nativeToken || chainMetadata_1.etherToken,
        rpcUrls: {
            public: { http: [metadata.rpcUrls[0].http] },
            default: { http: [metadata.rpcUrls[0].http] },
        },
        blockExplorers: metadata.blockExplorers?.length
            ? {
                default: {
                    name: metadata.blockExplorers[0].name,
                    url: metadata.blockExplorers[0].url,
                },
            }
            : undefined,
        testnet: !!metadata.isTestnet,
    };
}
exports.chainMetadataToWagmiChain = chainMetadataToWagmiChain;
