// @hyperlane-xyz/core/dist/factories/contracts/mock/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MockToken__factory = exports.MockPortalBridge__factory = exports.MockMailbox__factory = exports.MockHyperlaneEnvironment__factory = exports.MockCircleTokenMessenger__factory = exports.MockCircleMessageTransmitter__factory = exports.mockErc5164Sol = void 0;
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
exports.mockErc5164Sol = __importStar(require("./MockERC5164.sol"));
var MockCircleMessageTransmitter__factory_1 = require("./MockCircleMessageTransmitter__factory");
Object.defineProperty(exports, "MockCircleMessageTransmitter__factory", { enumerable: true, get: function () { return MockCircleMessageTransmitter__factory_1.MockCircleMessageTransmitter__factory; } });
var MockCircleTokenMessenger__factory_1 = require("./MockCircleTokenMessenger__factory");
Object.defineProperty(exports, "MockCircleTokenMessenger__factory", { enumerable: true, get: function () { return MockCircleTokenMessenger__factory_1.MockCircleTokenMessenger__factory; } });
var MockHyperlaneEnvironment__factory_1 = require("./MockHyperlaneEnvironment__factory");
Object.defineProperty(exports, "MockHyperlaneEnvironment__factory", { enumerable: true, get: function () { return MockHyperlaneEnvironment__factory_1.MockHyperlaneEnvironment__factory; } });
var MockMailbox__factory_1 = require("./MockMailbox__factory");
Object.defineProperty(exports, "MockMailbox__factory", { enumerable: true, get: function () { return MockMailbox__factory_1.MockMailbox__factory; } });
var MockPortalBridge__factory_1 = require("./MockPortalBridge__factory");
Object.defineProperty(exports, "MockPortalBridge__factory", { enumerable: true, get: function () { return MockPortalBridge__factory_1.MockPortalBridge__factory; } });
var MockToken__factory_1 = require("./MockToken__factory");
Object.defineProperty(exports, "MockToken__factory", { enumerable: true, get: function () { return MockToken__factory_1.MockToken__factory; } });
